lib/puppet/provider/ntp_config/cisco_nexus.rb

```diff
@@ -0,0 +1,81 @@
+require 'cisco_node_utils'
+
+module Puppet; end # rubocop:disable Style/Documentation
+module Puppet::ResourceApi
+  # Implementation for the ntp_config type using the Resource API.
+  class Puppet::Provider::NtpConfig::CiscoNexus
+    def canonicalize(_context, resources)
+      resources.each do |resource|
+        resource[:trusted_key] = resource[:trusted_key].sort_by(&:to_i).map(&:to_s) if resource[:trusted_key]
+      end
+      resources
+    end
+
+    def set(context, changes)
+      changes.each do |name, change|
+        is = change[:is]
+        should = change[:should]
+
+        if should != is
+          update(context, name, should)
+        end
+      end
+    end
+
+    def get(_context)
+      @ntp_config ||= Cisco::NtpConfig.new('default')
+
+      current_state = {
+        name:             'default',
+        authenticate:     @ntp_config.authenticate ? true : false,
+        source_interface: @ntp_config.source_interface ? @ntp_config.source_interface : 'unset',
+        trusted_key:      @ntp_config.trusted_key ? @ntp_config.trusted_key : ['unset'],
+      }
+
+      [current_state]
+    end
+
+    def update(context, name, should)
+      validate_should(should)
+      context.notice("Updating '#{name}' with #{should.inspect}")
+      @ntp_config ||= Cisco::NtpConfig.new('default')
+      @ntp_config.authenticate = should[:authenticate] unless should[:authenticate].nil?
+      @ntp_config.source_interface = should[:source_interface] == 'unset' ? nil : should[:source_interface]
+      handle_trusted_keys(should[:trusted_key]) if should[:trusted_key]
+    end
+
+    def handle_trusted_keys(should_trusted_keys)
+      @ntp_config ||= Cisco::NtpConfig.new('default')
+      if should_trusted_keys == ['unset']
+        remove = @ntp_config.trusted_key.map(&:to_s) if @ntp_config.trusted_key
+      elsif @ntp_config.trusted_key
+        # Otherwise calculate the delta
+        remove = @ntp_config.trusted_key.map(&:to_s).sort -
+                 should_trusted_keys.map(&:to_s).sort
+        remove.delete('unset')
+      end
+
+      remove.each do |key|
+        @ntp_config.trusted_key_set(false, key) unless key == 'unset'
+      end if remove
+      # Get array of keys to add
+      return if should_trusted_keys == ['unset']
+      if @ntp_config.trusted_key
+        add = should_trusted_keys.map(&:to_s).sort -
+              @ntp_config.trusted_key.map(&:to_s).sort
+      else
+        add = should_trusted_keys.map(&:to_s).sort
+      end
+      remove.delete('unset') if remove
+      add.each do |key|
+        @ntp_config.trusted_key_set(true, key)
+      end
+    end
+
+    def validate_should(should)
+      raise Puppet::ResourceError, 'Invalid name, `name` must be `default`' if should[:name] != 'default'
```

Comment:
Looking at https://github.com/puppetlabs/cisco-network-puppet-module/blob/89ca2c67aaa2e4841790086b1ee8825facffea5d/lib/puppet/provider/ntp_config/cisco.rb#L52 the old provider allowed `ntp_config`s that were not called `default`. What's happening here?

Reply:
From node-utils, it will only even return `ntp_configs` that are `default`. I can change it to accept anything, but here is the code from node-utils:

```
    def self.ntpconfigs
      hash = {}
      hash['default'] = NtpConfig.new('default')
      hash
    end
```

Reply:
:man_facepalming: :back_away:

Reply:
In general we should not be duplicating validation or constraints across nodeutils/providers. I'd say nodeutils is the place to verify stuff, as it is the more widely shared codebase.

Reply:
Happy to refactor this then and make the provider less constraining. Was just worried about performance really as making the validation at the provider is a few milliseconds, whereas the call to node-utils does a lot in the background such as fetching the device config and parsing it. 🔥

Reply:
Looks like the nodeutils does perform this validation https://github.com/cisco/cisco-network-node-utils/blob/4e0e66d76dded6aa437993fd04748dde3db6ffaa/lib/cisco_node_utils/ntp_config.rb#L29